content/advent-2018/go-arrow.md

```diff
@@ -0,0 +1,534 @@
+		},
+		nil, // no metadata
+	)
+
+	b := array.NewRecordBuilder(mem, schema)
+	defer b.Release()
+
+	b.Field(0).(*array.Int32Builder).AppendValues(
+		[]int32{1, 2, 3, 4, 5, 6},
+		nil,
+	)
+	b.Field(0).(*array.Int32Builder).AppendValues(
+		[]int32{7, 8, 9, 10},
+		[]bool{true, true, false, true},
+	)
+	b.Field(1).(*array.Float64Builder).AppendValues(
+		[]float64{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, 
+		nil,
+	)
+
+	rec1 := b.NewRecord()
+	defer rec1.Release()
+
+	b.Field(0).(*array.Int32Builder).AppendValues([]int32{
+		11, 12, 13, 14, 15, 16, 17, 18, 19, 20}, 
+		nil,
+	)
+	b.Field(1).(*array.Float64Builder).AppendValues([]float64{
+		11, 12, 13, 14, 15, 16, 17, 18, 19, 20}, 
+		nil,
+	)
+
+	rec2 := b.NewRecord()
+	defer rec2.Release()
+
+	tbl := array.NewTableFromRecords(schema, []array.Record{rec1, rec2})
+	defer tbl.Release()
+
+	tr := array.NewTableReader(tbl, 5)
+	defer tr.Release()
+
+	n := 0
+	for tr.Next() {
+		rec := tr.Record()
+		for i, col := range rec.Columns() {
+			fmt.Printf("rec[%d][%q]: %v\n", n, rec.ColumnName(i), col)
+		}
+		n++
+	}
+}
+```
+
+Executing the code above would result in:
+```
+$> go run ./table-reader.go
+rec[0]["f1-i32"]: [1 2 3 4 5]
+rec[0]["f2-f64"]: [1 2 3 4 5]
+rec[1]["f1-i32"]: [6 7 8 (null) 10]
+rec[1]["f2-f64"]: [6 7 8 9 10]
+rec[2]["f1-i32"]: [11 12 13 14 15]
+rec[2]["f2-f64"]: [11 12 13 14 15]
+rec[3]["f1-i32"]: [16 17 18 19 20]
+rec[3]["f2-f64"]: [16 17 18 19 20]
+```
+
+## CSV data
+
+Finally, to conclude with our quick whirlwind overview of what Go Arrow can provide now, let us consider the [arrow/csv](https://godoc.org/github.com/apache/arrow/go/arrow/csv) package.
+As mentioned earlier, many analysis pipelines start with ingesting [CSV](https://en.wikipedia.org/wiki/Comma-separated_values).
+The Go standard library already provides a package to decode and encode data using this comma-separated values (CSV) "format".
+But what the [encoding/csv](https://godoc.org/encoding/csv) package exposes is slices of `string`s, not values of a specific type.
+This is left to the user of that package.
+
+The [arrow/csv](https://godoc.org/github.com/apache/arrow/go/arrow/csv) package leverages the `arrow.Schema` and `array.Record` types to provide a typed and (eventually) scalable+optimized API.
+
+```go
+package main
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/apache/arrow/go/arrow"
+	"github.com/apache/arrow/go/arrow/csv"
+)
+
+func main() {
+	f := bytes.NewBufferString(`## a simple set of data: int64;float64;string
+0;0;str-0
+1;1;str-1
+2;2;str-2
+3;3;str-3
+4;4;str-4
+5;5;str-5
+6;6;str-6
+7;7;str-7
+8;8;str-8
+9;9;str-9
+`)
+
+	schema := arrow.NewSchema(
+		[]arrow.Field{
+			{Name: "i64", Type: arrow.PrimitiveTypes.Int64},
+			{Name: "f64", Type: arrow.PrimitiveTypes.Float64},
+			{Name: "str", Type: arrow.BinaryTypes.String},
+		},
+		nil, // no metadata
+	)
+	r := csv.NewReader(
+		f, schema,
+		csv.WithComment('#'), csv.WithComma(';'),
+		csv.WithChunk(3),
```

Comment:
I find the usage of `WithChunk` to be kind of confusing, along with the "diagonal" iteration of the data a few lines down. Is there any particular aspect of arrow that's being illustrated by doing it thing way, as opposed to a standard 2-d row-by-column iteration? It seems more obvious what's going on if the chunk size is not equal to the number of columns, like 4 or something. Otherwise, I think it would be better to go with code like this:

```go
package main
import (
	"bytes"
	"fmt"
	"github.com/apache/arrow/go/arrow"
	"github.com/apache/arrow/go/arrow/csv"
)
 func main() {
	f := bytes.NewBufferString(`## a simple set of data: int64;float64;string
0;0.0;str-0
1;0.1;str-1
2;0.2;str-2
3;0.3;str-3
4;0.4;str-4
5;0.5;str-5
6;0.6;str-6
7;0.7;str-7
8;0.8;str-8
9;0.9;str-9
`)
	schema := arrow.NewSchema(
		[]arrow.Field{
			{Name: "i64", Type: arrow.PrimitiveTypes.Int64},
			{Name: "f64", Type: arrow.PrimitiveTypes.Float64},
			{Name: "str", Type: arrow.BinaryTypes.String},
		},
		nil, // no metadata
	)
	r := csv.NewReader(
		f, schema,
		csv.WithComment('#'), csv.WithComma(';'),
	)
	defer r.Release()
	n := 0
	for r.Next() {
		rec := r.Record()
		for i, col := range rec.Columns() {
			fmt.Printf("rec[%d][%q]: %v\n", n, rec.ColumnName(i), col)
		}
		n++
	}
}
```

```
$> go run ./read-csv.go
rec[0]["i64"]: [0]
rec[0]["f64"]: [0]
rec[0]["str"]: ["str-0"]
rec[1]["i64"]: [1]
rec[1]["f64"]: [0.1]
rec[1]["str"]: ["str-1"]
rec[2]["i64"]: [2]
rec[2]["f64"]: [0.2]
rec[2]["str"]: ["str-2"]
rec[3]["i64"]: [3]
rec[3]["f64"]: [0.3]
rec[3]["str"]: ["str-3"]
rec[4]["i64"]: [4]
rec[4]["f64"]: [0.4]
rec[4]["str"]: ["str-4"]
rec[5]["i64"]: [5]
rec[5]["f64"]: [0.5]
rec[5]["str"]: ["str-5"]
rec[6]["i64"]: [6]
rec[6]["f64"]: [0.6]
rec[6]["str"]: ["str-6"]
rec[7]["i64"]: [7]
rec[7]["f64"]: [0.7]
rec[7]["str"]: ["str-7"]
rec[8]["i64"]: [8]
rec[8]["f64"]: [0.8]
rec[8]["str"]: ["str-8"]
rec[9]["i64"]: [9]
rec[9]["f64"]: [0.9]
rec[9]["str"]: ["str-9"]
```

Reply:
good point about the "unlucky" or confusing value of `3` for the chunking.
I'll replace with `4`.

thanks!